src/nupic/ntypes/Sdr.hpp

```diff
@@ -0,0 +1,273 @@
+     *
+     * @param value An SDR to replicate.
+     */
+    SparseDistributedRepresentation(const SparseDistributedRepresentation value);
+
+    /**
+     * @returns A list of dimensions of the SDR.
+     */
+    const vector<UInt> getDimensions() const
+        {return dimensions;};
+
+    /**
+     * @returns The total number of boolean values in the SDR.
+     */
+    const UInt getSize() const
+        {return size;};
+
+    /**
+     * Remove the value from this SDR.  Attempting to get the value of an empty
+     * SDR will raise an exception.
+     */
+    void clear();
+
+    /**
+     * Set all of the values in the SDR to false.  This method overwrites the
+     * SDRs current value.
+     */
+    void zero();
+
+    /**
+     * Assigns a new value to the SDR, overwritting the current value.
+     *
+     * @param value A dense array in a vector<char> to assign to the SDR.
+     */
+    void setDense( const vector<Byte> value );
+
+    /**
+     * Assigns a new value to the SDR, overwritting the current value.
+     *
+     * @param value A dense C-style array of UInt's to assign to the SDR.  The
+     * given value must have as many elements as this SDR, as given by 
+     * SDR.getSize().
+     */
+    void setDense( const UInt *value );
+
+    /**
+     * Assigns a new value to the SDR, overwritting the current value.
+     *
+     * @param value A dense byte Array to assign to the SDR.
+     */
+    void setDense( const ArrayBase *value );
+
+    /**
+     * Assigns a vector of sparse indices to the SDR.  These indicies are into
+     * the flattened SDR space.  This overwrites the SDR's current value.
+     *
+     * @param value A vector of indices to assign to the SDR.
+     */
+    void setFlatIndex( const vector<UInt> value );
+
+    /**
+     * Assigns an array of sparse indices to the SDR.  These indicies are into
+     * the flattened SDR space.  This overwrites the SDR's current value.
+     *
+     * @param value A C-style array of indices to assign to the SDR.
+     * @param num_values The number of elements in the 'value' array.
+     */
+    void setFlatIndex( const UInt *value, const UInt num_values );
+
+    /**
+     * Assign a list of indices to the SDR, overwritting the SDRs current value.
+     * These indices are into the SDR space with dimensions.  The outter list
+     * is indexed using an index into the dimensions list, see
+     * SDR.getDimensions().  The inner lists are indexed in parallel, they
+     * contain the coordinates of the true values in the SDR.
+     *
+     * @param value A list of lists containing the coordinates of the true
+     * values to assign to the SDR.
+     */
+    void setIndex( const vector<vector<UInt>> value );
+
+    /**
+     * Assigns a deep copy of the given SDR to this SDR.  This overwrites the
+     * current value of this SDR.
+     *
+     * @param value An SDR to copy the value of.
+     */
+    void assign( const SparseDistributedRepresentation value );
```

Comment:
Both shallow and deep copies are only two operations that might be performed on the buffer.  There are all of the boolean operators. Dot product. Sum, and many others.  If you look at numpy functions you can get some ideas but obviously we do not need all of them.  In the Python code the SDR is a numpy array so when porting to C++ it would be nice to have some of those functions built into the SDR object to simplify porting.

Reply:
First, there are no shallow copies of SDRs, they *should* all be deep copies.  If you want a shallow copy you should get a pointer to the SDR.

Second, for all of the per-entry functions which numpy has so many of: I'd rather not implement stuff which isn't specific to SDRs.   I will add "overlap" and "sum", but not a "dot product" unless you make the case that dot-product would be useful for SDRs and difficult to do without a custom SDR-based implementation.

Reply:
`I'd rather not implement stuff which isn't specific to SDRs
`
I absolutely agree.

` there are no shallow copies of SDRs`
I am trying to visualize how this would be used.  region A outputs into a buffer provided by an SDR object.  We want to pass that to region B which is going to get that same pointer from the SDR object and read that data (using whatever indexing it might want to use).  How do we get that SDR from region A to a place where region B can access it without making a copy of the underlining array of data.

Reply:
These shallow copies complicates things a bit.  Here is my idea of how to implement:

Split the SDR into two classes, a public SDR class with dimensions, and a private class with the rest of the data.  Multiple SDR classes can point to a data class.  

```
class SDR {
    class SDR_Data {
         ...
    }

    vector<UInt> dimensions;
    std::shared_ptr<SDR_Data> data;
}
```

I've never used shared_ptr, is this a good way to go?
Would this do everything needed for the regions to work?